packages/design-tokens-schema/src/basis-tokens.ts

```diff
@@ -300,19 +320,26 @@ export const StrictThemeSchema = ThemeSchema.transform(addContrastExtensions)
     }
 
     // Validation 2: Check that colors have sufficient contrast
-    walkColors(root, (foreground, path) => {
-      if (!Array.isArray(foreground.$extensions?.[EXTENSION_CONTRAST_WITH])) return;
+    walkColors(root, (token, path) => {
+      if (!Array.isArray(token.$extensions?.[EXTENSION_CONTRAST_WITH])) return;
+
+      const comparisons = token.$extensions[EXTENSION_CONTRAST_WITH];
+      const baseColor = getActualValue<ColorValue>(token);
 
-      for (const { color: background, ratio: expectedContrast } of foreground.$extensions[EXTENSION_CONTRAST_WITH]) {
-        const contrast = compareContrast(foreground, background);
-        const colorRefName = background.$extensions?.[EXTENSION_RESOLVED_FROM];
-        if (contrast < expectedContrast) {
+      for (const { color: background, expectedRatio } of comparisons) {
+        const compareColor = getActualValue<ColorValue>(background);
+        const contrast = compareContrast(baseColor, compareColor);
+        const resolvedFrom = background.$extensions?.[EXTENSION_RESOLVED_FROM];
+        const colorRefName = background.$extensions?.[EXTENSION_RESOLVED_AS]
+          ? `${resolvedFrom} -> ${background.$value}`
+          : resolvedFrom;
```

Comment:
I think my misunderstanding stems from the following question; does every token with a token ref have the `EXTENSION_RESOLVED_FROM`? My assumption was it would not, but I guess it does?

And that makes the `${resolvedFrom} -> ${background.$value}` collapse the whole chain, ie the former is the ref that holds the actual value, and the latter is first token in the chain of references?

If so, _very_ nice

Reply:
~~If the ref was valid, then yes.~~ Only if we automatically added a CONTRAST_WITH extension, that's the only place where we set the RESOLVED_FROM extension.